lib/auth/predicate/builtins.go

```
@@ -0,0 +1,361 @@
+)
+
+func getIdentifier(obj any, selectors []string) (any, error) {
+	switch selectors[0] {
+	case "false":
+		return false, nil
+	case "true":
+		return true, nil
+	}
+
+	for _, s := range selectors {
+		if obj == nil || reflect.ValueOf(obj).IsNil() {
+			return nil, trace.BadParameter("cannot take field of nil")
+		}
+
+		if m, ok := obj.(map[string]any); ok {
+			obj = m[s]
+			continue
+		}
+
+		val := reflect.ValueOf(obj)
+		ty := reflect.TypeOf(obj)
+		if ty.Kind() == reflect.Interface || ty.Kind() == reflect.Ptr {
+			val = reflect.ValueOf(obj).Elem()
+			ty = val.Type()
+		}
+
+		if ty.Kind() == reflect.Struct {
+			for i := 0; i < ty.NumField(); i++ {
+				tagValue := ty.Field(i).Tag.Get("json")
+				parts := strings.Split(tagValue, ",")
+				if parts[0] == s {
+					obj = val.Field(i).Interface()
+					break
+				}
+			}
+
+			continue
+		}
+
+		return nil, trace.BadParameter("cannot take field of type: %T", obj)
+	}
+
+	return obj, nil
+}
+
+func getProperty(m any, k any) (any, error) {
+	switch mT := m.(type) {
+	case map[string]any:
+		kS, ok := k.(string)
+		if !ok {
+			return nil, trace.BadParameter("unsupported key type: %T", k)
+		}
+
+		return mT[kS], nil
+	default:
+		return nil, trace.BadParameter("cannot take property of type: %T", m)
+	}
+}
+
+func builtinOpAnd(a, b bool) any {
+	return a && b
+}
+
+func builtinOpOr(a, b bool) any {
+	return a || b
+}
+
+func builtinOpNot(a bool) any {
+	return !a
+}
+
+func builtinOpEquals(a, b any) any {
+	return reflect.DeepEqual(a, b)
+}
+
+func builtinOpLT(a, b any) any {
+	if reflect.TypeOf(a) != reflect.TypeOf(b) {
+		return false
+	}
+
+	switch aT := a.(type) {
+	case string:
+		return aT < b.(string)
+	case int:
+		return aT < b.(int)
+	case float32:
+		return aT < b.(float32)
+	default:
+		return false
+	}
+}
+
+func builtinOpGT(a, b any) any {
+	return builtinOpLT(b, a)
+}
+
+func builtinOpLE(a, b any) any {
+	if reflect.TypeOf(a) != reflect.TypeOf(b) {
+		return false
+	}
```

I think you can return an error here as well